clean architecture.md

```diff
@@ -0,0 +1,144 @@
+
+### Folder structure
+- domain
+  - interfaces
+  - entities
+- adaptors
+- routers
+- glue
+
+### Entities
+The entities are respresentations of the different data structures used in the system. 
+
+```
+export interface Talent {
+  name: string;
+  skillLevel: number;
+}
+```
+
+### Interfaces
+The interfaces describe the various inputs and outputs expected by the domain functions, including the higher order domain functions. We could have used typescript interfaces but instead opted for types, which we thought were simpler to use and understand.
+
+```
+export type FunMaker = (
+  relationships: Relationship[],
+  environment: Environment,
+  money?: number
+) => Fun;
+```
+
+### Domain
+Domain functions contain the business logic, of which we had two types: higher order functions and simple functions.
+
+#### Simple functions
+The simple functions are functions, usually pure, that take an input and produce an output.
+
+```
+const calculateTotal = ({ shipping: number, taxes: number, total: number }) => {
+  return shipping + taxes + total;
+};
+```
+
+#### Higher order functions
+For the functions that include orchestration logic implementing different adaptors, we use higher order functions that accept all the necessary dependencies and return a function that implement those dependencies.
+
+```
+const buildHappinessGenerator = (
+ makeMoney: MoneyMaker,
+ makeRelationships: RelationshipMaker,
+ makeFun: FunMaker,
+ determinePurpose: PurposeDeterminer,
+ rageAgainstNihilism: NihilismRager,
+ seekHelp: HelpSeeker
+) => {
+  return (
+    values: Value[], 
+    talents: Talent[],
+    weaknesses: Weakness[],
+    environment: Environment
+  ): Happiness => {
+    try {
+      const purpose = determinePurpose(values, talents, weaknesses, environment);
+      const relationships = makeRelationships(values, environment);
+      const money = makeMoney(talents, environment, purpose);
+      const fun = makeFun(relationships, environment, money);
+
+      return rageAgainstNihilism(purpose, relationships, money, fun);
+    } catch (err) {
+      seekHelp(relationships);
+    }
+  };
+}
+```
+
+These are all expected to have tests and this is where the real power is revealed. Whereas in the past we’d have to use nock to intercept HTTP requests and create a whole bunch of fixture files, now we could create mock functions representing the injected dependencies, focusing our testing efforts on the logic and orchestration.
+
+### Adaptors
+Adaptors are functions that encapsulate I/O, whether that is reading/writing from a data store, calling an API, shooting a firework into the air, or whatever silly example you can think of. These functions are meant to be very simple, with no business logic, and as a result, don’t need tests and should be excluded from code coverage.
```

Comment:
I would probably state adaptors don't need _unit_ tests, because they wouldn't really prove anything. To really test adaptors would require calling the actual API/DB/fireworks/whatever, which would be outside the realm of unit tests and probably more into the realm of automated QE tests.

Reply:
Thanks for the suggestion. That's sounds like an important distinction to make